File: 1605-minimum-number-of-days-to-make-m-bouquets/minimum-number-of-days-to-make-m-bouquets.cpp
class Solution {
public:
    bool check(vector<int>& bloomDay, int m, int k, int mid)
    {
        int c = 0;
        for(auto it : bloomDay){
            if(it <= mid){
                c++;
                if(c == k){
                    m--;
                    c = 0;
                }
            }
            else c = 0;
        }
        return m <= 0;
    }

    int minDays(vector<int>& bloomDay, int m, int k) 
    {
        int n = bloomDay.size();
        if(m > n/k) return -1; // checking (m*k > n), use divide to avoid integer overflow

        int lo = *min_element(bloomDay.begin(), bloomDay.end());
        int hi = *max_element(bloomDay.begin(), bloomDay.end());
        int ans = hi;

        while(lo <= hi)
        {
            int mid = lo + (hi-lo)/2;

            if(check(bloomDay, m, k, mid))
            {
                ans = min(ans, mid);
                hi = mid - 1;
            }
            else lo = mid + 1;
        }

        return ans;
    }
};

File: 1605-minimum-number-of-days-to-make-m-bouquets/minimum-number-of-days-to-make-m-bouquets.py
class Solution:
    def minDays(self, bloomDay: List[int], m: int, k: int) -> int:
        def check(mid):
            bCounter = 0
            current = 0
            for day in bloomDay:
                if day <= mid:
                    current += 1
                    if current == k:
                        current = 0
                        bCounter += 1
                else:
                    current = 0
                
            return bCounter >= m

        lo = 1
        hi = max(bloomDay)
        res = -1

        while lo <= hi:
            mid = (lo + hi)//2

            if check(mid):
                res = mid
                hi = mid - 1
            else:
                lo = mid + 1

        return res